### PR TITLE
Refactor DartdocConfig option usage, part 4 of 4

### LINKED ISSUE #1609: Allow running from Dart code.
For the `dartdocs.org` site, it would be better if we could run `dartdoc` without executing a new process. The current `bin/dartdoc.dart` does a lot of heavy lifting, and also one-off environment setup, and copying that is not straightforward.

Is there a way to simply call a method to run `dartdoc` repeatedly from the Dart code?

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #1609.  And finally, the payoff.  Eliminate the old dartdoc config classes, seriously clean up `bin/dartdoc.dart`, define options a little more closely to where they are used, and make it at least not completely arcane to build your own generators in dartdoc extensions.

This implements, but doesn't finalize #1674.  I'm going to write a series of integration tests for each new option and promote them to supported once I'm convinced they work reliably from config files, but that's an awful lot to throw into this PR, too.

### CODE REVIEW

File: lib/src/model.dart [hunk 1]
```diff
@@ -3845,15 +3851,15 @@ class PackageGraph extends Canonicalization
...
+  final DartdocOptionContext config;
 
   Map<String, Map<String, dynamic>> __crossdartJson;
   // TODO(jcollins-g): move to [Package]
   Map<String, Map<String, dynamic>> get crossdartJson {
     if (__crossdartJson == null) {
       // TODO(jcollins-g): allow crossdart.json location to be configurable
       var crossdartFile =
-          new File(pathLib.join(config.inputDir.path, "crossdart.json"));
+          new File(pathLib.join(config.inputDir, "crossdart.json"));
```
Comment (pq) [inline lib/src/model.dart]: => 'crossdart.json'

File: lib/src/model.dart [hunk 2]
```diff
@@ -5000,7 +5007,7 @@ abstract class SourceCodeMixin implements Documentable {
         }
       } else if (uri.startsWith("dart:")) {
```
Comment (pq) [inline lib/src/model.dart]: Single quotes?

File: lib/src/model.dart [hunk 3]
```diff
@@ -5000,7 +5007,7 @@ abstract class SourceCodeMixin implements Documentable {
...
         var packageName = "sdk";
-        var packageVersion = config.sdkVersion;
```
Comment (pq) [inline lib/src/model.dart]: And here?

File: lib/src/model.dart [hunk 4]
```diff
@@ -5237,31 +5244,34 @@ class TypeParameter extends ModelElement {
...
-          PhysicalResourceProvider.INSTANCE.getFolder(config.sdkDir.path));
+          PhysicalResourceProvider.INSTANCE.getFolder(config.sdkDir));
     }
     return _sdk;
   }
 
   EmbedderSdk _embedderSdk;
   EmbedderSdk get embedderSdk {
-    if (_embedderSdk == null && packageMeta.isSdk == false) {
+    if (_embedderSdk == null && config.packageMeta.isSdk == false) {
```
Comment (pq) [inline lib/src/model.dart]: `!config.packageMeta.isSdk` ?

File: lib/src/model.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart (hunk 2 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart (hunk 3 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart (hunk 4 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

**Pull request merged**